src/libexpr/primops.cc

```diff
@@ -1043,6 +1043,53 @@ static RegisterPrimOp primop_trace({
     .fun = prim_trace,
 });
 
+static void prim_warn(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    state.forceValue(*args[0], pos);
+
+    {
+        BaseError msg(args[0]->type() == nString
+                ? std::string(args[0]->string_view())
+                : ({
+                    std::stringstream s;
+                    s << ValuePrinter(state, *args[0]);
+                    s.str();
+                }));
+        msg.atPos(state.positions[pos]);
+        auto info = msg.info();
+        info.level = lvlWarn;
+        logWarning(info);
+    }
+
+    if (evalSettings.builtinsAbortOnWarn) {
+        state.error<Abort>("aborting to reveal stack trace of warning, as abort-on-warn is set").debugThrow();
```

Comment:
Optional: we could also just throw our `BaseError` with this as extra context.

Reply:
I prefer to treat them separately, because they don't have the same message, and I don't think it should be just a single message.

Reply:
Totally fine! Though just to be clear, I was thinking the this message would become an `addTrace` call --- I too don't think it should be a single message.